### PR TITLE
Add support for custom binding

### PR DESCRIPTION
**Pull request by Flanker32**
Add support for custom binding, for issue #559

### CODE REVIEW

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/bindings/CustomBinding.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * license information.
+ */
+
+package com.microsoft.azure.maven.function.bindings;
+
+import java.lang.annotation.Annotation;
+
+public class CustomBinding extends Binding {
+
+    private com.microsoft.azure.functions.annotation.CustomBinding customBindingAnnotation;
```
Comment (andxu) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/bindings/CustomBinding.java]: what's the reason for using full package name here? com.microsoft.azure.functions.annotation.CustomBinding

Comment (Flanker32) [reply azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/bindings/CustomBinding.java]: This is the annotation class from azure function java library, in order not to mix it with CustomBinding Object, use the full name.

**Comment by Flanker32** [comment]
@pragnagopa  Sorry for reply late, cases has been updated

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/handlers/AnnotationHandlerImplTest.java [hunk 2]
```diff
@@ -87,7 +94,24 @@
...
+
+        @FunctionName(CUSTOM_BINDING_FUNCTION)
+        public void customBindingMethod(
+            @TestCustomBinding(index = "testIndex", path = "testPath") String customTriggerInput) {
+        }
+
+        @Target(ElementType.PARAMETER)
+        @Retention(RetentionPolicy.RUNTIME)
+        @CustomBinding(direction = "in", name = "message", type = "customBinding")
+        public @interface TestCustomBinding {
```
Comment (pragnagopa) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/handlers/AnnotationHandlerImplTest.java]: Using CustomBinding by itself is supported as well. Please add test.

**Comment by pragnagopa** [comment]
@Flanker32 - we should also support
```java
@Target(ElementType.PARAMETER)
@Retention(RetentionPolicy.RUNTIME)
@CustomBinding(direction = "in", name = "", type = "customBinding")
public @interface TestCustomBinding {
   String index();
   String path();
   String name();
}
```
In this case, name() on TestCustomBinding should override name given in customBinding

**Comment by jdneo** [comment]
@pragnagopa It's a good idea that using CustomBinding to extends the binding model. But may I ask in what kind of scenario, the user needs to override the `name`, `direction` or `type` fields defined in `@CustomBinding`. So far, I didn't see the necessity for this kind of usage.

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/bindings/Binding.java [hunk 3]
```diff
@@ -100,4 +100,21 @@ public String toString() {
                 .append(" ]")
                 .toString();
     }
+
+    protected String getDirectionFromAttribute() {
+        if (this.direction != null) {
```
Comment (Flanker32) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/bindings/Binding.java]: We may split this function, do the null check of direction in `getDirection` and only get direction from attribute map here.

**Comment by jdneo** [comment]
@pragnagopa Could you please take a look?

**Comment by anthonychu** [comment]
> @pragnagopa It's a good idea that using CustomBinding to extends the binding model. But may I ask in what kind of scenario, the user needs to override the `name`, `direction` or `type` fields defined in `@CustomBinding`. So far, I didn't see the necessity for this kind of usage.

@jdneo The main scenario is to use `CustomBinding` to define annotations to be used in multiple places (therefore it is useful to control the name where the annotation is used). We've [been asked](https://github.com/Azure/azure-functions-java-library/pull/78#pullrequestreview-203504473) to use `CustomBinding` to build annotations for SignalR Service. It's not possible to do this without support for overriding the name (overriding `direction` and `type` are not necessary), so this is currently a blocker for us.

Here's an example:

```java
@Target({ElementType.PARAMETER, ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
@CustomBinding(direction = "out", name = "", type = "SignalR")
public @interface SignalR {
   String name();
   String hubName();
}
```

Usage:

```java
@FunctionName("messages")
public void sendMessage(
        @HttpTrigger(
            name = "req", 
            methods = { HttpMethod.POST },
            authLevel = AuthorizationLevel.ANONYMOUS) 
            HttpRequestMessage<Object> req,
        @SignalR(name = "chatMessage", hubName = "chat") OutputBinding<SignalRMessage> chatMessage,
        @SignalR(name = "notificationMessage", hubName = "notification") OutputBinding<SignalRMessage> notificationMessage) {

    chatMessage.setValue(/* a chat message */);
    notificationMessage.setValue(/* a notification message */);
}
```

/cc @JialinXin @chenkennt @sffamily

**Comment by jdneo** [comment]
@anthonychu Thank you for clarifying. Get your points now!

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/bindings/BindingEnum.java [hunk 4]
```diff
@@ -32,7 +32,8 @@
     TableOutput("table", Direction.OUT, true),
     TimerTrigger("timerTrigger", Direction.IN),
     TwilioSmsOutput("twilioSms", Direction.OUT),
-    CustomBinding("customBinding", null);
+    CustomBinding("customBinding", null),
+    UnknownBinding(null, null);
```
Comment (Flanker32) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/bindings/BindingEnum.java]: I think ExtendedCustomBindingEnum could be better

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/bindings/Binding.java [hunk 5]
```diff
@@ -101,20 +102,22 @@ public String toString() {
                 .toString();
     }
 
-    protected String getDirectionFromAttribute() {
-        if (this.direction != null) {
-            return direction.toString();
+    protected void addProperties(List<String> requiredAttributes, Object value, Method method) {
```
Comment (Flanker32) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/bindings/Binding.java]: I think we don't need to pass requiredAttributes as it is an member properties of Binding class

**Comment by jdneo** [comment]
Currently, the plugin should be able to resolve `CustomBinding` and the annotations which are extended from `CustomBinding`. While from the implementation, we found that the logic needs to be refactored.

Considering Java GA is coming soon, we decide to put the refactoring work after GA. Tracking issue: #583

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/handlers/AnnotationHandlerImplTest.java [hunk 6]
```diff
@@ -87,7 +96,33 @@
...
+        }
+
+        @FunctionName(CUSTOM_BINDING_FUNCTION)
+        public void customBindingMethod(
+                @CustomBinding(name = "input", type = "customBinding", direction = "in") String input) {
+        }
+
+        @FunctionName(EXTENDING_CUSTOM_BINDING_FUNCTION)
+        public void extendingcustomBindingMethod(
+            @TestCustomBinding(direction = "out", name = "extendingCustomBinding", path = "testPath")
```
Comment (pragnagopa) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/handlers/AnnotationHandlerImplTest.java]: direction cannot be overridden as the annotation author decides the direction for a given binding type.

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/handlers/AnnotationHandlerImplTest.java [hunk 7]
```diff
@@ -87,7 +96,33 @@
...
+        @Retention(RetentionPolicy.RUNTIME)
+        @CustomBinding(direction = "in", name = "message", type = "customBinding")
+        public @interface TestCustomBinding {
+            String direction();
+            String path();
+            String name();
+        }
+
+        @FunctionName(CUSTOM_BINDING_FUNCTION)
+        public void customBindingMethod(
```
Comment (pragnagopa) [inline azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/handlers/AnnotationHandlerImplTest.java]: Add one more test where TestCustomBinding does not define a name.
```java
@FunctionName(EXTENDING_CUSTOM_BINDING_FUNCTION)
        public void extendingcustomBindingMethod2(
            @TestCustomBinding(direction = "out",  path = "testPath")
                    String customTriggerInput) {
        }
```
In this case, name should be taken from customBinding: message

**Comment by anthonychu** [comment]
I just tested this out as well. It works great. Thanks for the quick work on this @jdneo

Updated this sample to use it:
https://github.com/Azure-Samples/signalr-service-quickstart-serverless-chat/blob/antchu/add-java/src/chat/java/src/main/java/com/function/Functions.java

File: azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/handlers/AnnotationHandlerImplTest.java (hunk 6 above)
Comment (jdneo) [reply azure-functions-maven-plugin/src/test/java/com/microsoft/azure/maven/function/handlers/AnnotationHandlerImplTest.java]: Does that mean only name can be overridden?

**Comment by jdneo** [comment]
@anthonychu Glad to know it works. I just make some tiny changes to the code to address the comment. Would you mind to try it again to make sure things go well?

@pragnagopa Regarding this PR, We found some problems. Could you please take a look?
- https://github.com/Azure/azure-functions-java-library/issues/83
- https://github.com/Azure/azure-functions-java-library/issues/69

**Pull request merged**